File: src/main/java/com/dndweapons/DndWeaponsMod.java
package com.dndweapons;

import com.dndweapons.catalog.Weapons;
import com.dndweapons.registry.WeaponRegistrar;
import com.dndweapons.registry.WeaponRegistrarImpl;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup;
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemStack;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class DndWeaponsMod implements ModInitializer {
    public static final String MOD_ID = "dndweapons";
    public static final Logger LOGGER = LoggerFactory.getLogger(MOD_ID);

    public static final RegistryKey<ItemGroup> CREATIVE_TAB =
        RegistryKey.of(RegistryKeys.ITEM_GROUP, Identifier.of(MOD_ID, "main"));

    @Override
    public void onInitialize() {
        LOGGER.info("DnD Weapons initializing...");

        WeaponRegistrar registrar = new WeaponRegistrarImpl();
        registrar.registerAll(Weapons.ALL);

        Registry.register(Registries.ITEM_GROUP, CREATIVE_TAB,
            FabricItemGroup.builder()
                .displayName(Text.translatable("itemGroup.dndweapons.main"))
                .icon(() -> new ItemStack(Registries.ITEM.get(Identifier.of(MOD_ID, "longsword"))))
                .build()
        );

        ItemGroupEvents.modifyEntriesEvent(CREATIVE_TAB).register(entries -> {
            for (var spec : Weapons.ALL) {
                if (spec.isVanillaMapped()) continue;
                var item = Registries.ITEM.get(Identifier.of(MOD_ID, spec.id()));
                if (item != null) entries.add(item);
            }
        });

        LOGGER.info("DnD Weapons initialized with {} weapons.", Weapons.ALL.size());
    }
}

File: src/main/kotlin/com/dndweapons/DndWeaponsMod.kt
package com.dndweapons

import com.dndweapons.catalog.Weapons
import com.dndweapons.registry.WeaponRegistrarImpl
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceKey
//? if <1.21.11 {
import net.minecraft.resources.ResourceLocation
//?}
//? if >=1.21.11 {
/*import net.minecraft.resources.Identifier as ResourceLocation
*///?}
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.ItemStack
import org.slf4j.Logger
import org.slf4j.LoggerFactory

private fun rl(namespace: String, path: String): ResourceLocation {
    //? if >=1.21 {
    return ResourceLocation.fromNamespaceAndPath(namespace, path)
    //?} else {
    /*return ResourceLocation(namespace, path)
    *///?}
}

object DndWeaponsMod : ModInitializer {

    const val MOD_ID: String = "dndweapons"
    val LOGGER: Logger = LoggerFactory.getLogger(MOD_ID)

    val CREATIVE_TAB: ResourceKey<CreativeModeTab> =
        ResourceKey.create(Registries.CREATIVE_MODE_TAB, rl(MOD_ID, "main"))

    override fun onInitialize() {
        LOGGER.info("DnD Weapons initializing...")

        val registrar = WeaponRegistrarImpl()
        registrar.registerAll(Weapons.ALL)

        Registry.register(
            BuiltInRegistries.CREATIVE_MODE_TAB, CREATIVE_TAB,
            FabricItemGroup.builder()
                .title(Component.translatable("itemGroup.dndweapons.main"))
                .icon { iconStack() }
                .build(),
        )

        ItemGroupEvents.modifyEntriesEvent(CREATIVE_TAB).register { entries ->
            for (spec in Weapons.ALL) {
                if (spec.isVanillaMapped) continue
                addToEntries(entries, spec)
            }
        }

        LOGGER.info("DnD Weapons initialized with {} weapons.", Weapons.ALL.size)
    }

    private fun iconStack(): ItemStack {
        //? if >=1.21.2 {
        return BuiltInRegistries.ITEM
            .get(rl(MOD_ID, "longsword"))
            .map { ItemStack(it) }
            .orElse(ItemStack.EMPTY)
        //?} else {
        /*return ItemStack(BuiltInRegistries.ITEM.get(rl(MOD_ID, "longsword")))
        *///?}
    }

    private fun addToEntries(entries: net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroupEntries, spec: com.dndweapons.catalog.WeaponSpec) {
        val itemId = rl(MOD_ID, spec.id)
        //? if >=1.21.2 {
        BuiltInRegistries.ITEM.get(itemId).ifPresent { entries.accept(it.value()) }
        //?} else {

        /*val item = BuiltInRegistries.ITEM.get(itemId)
        if (item != null) entries.accept(item)
        *///?}
    }
}
